Code/GraphMol/FilterCatalog/FunctionalGroupHierarchy.cpp

```diff
@@ -0,0 +1,195 @@
+  
+  {0,"Halogen","[$([F,Cl,Br,I]-!@[#6]);!$([F,Cl,Br,I]-!@C-!@[F,Cl,Br,I]);!$([F,Cl,Br,I]-[C,S](=[O,S,N]))]","Halogen",0},
+  {1,"Halogen.Aromatic","[F,Cl,Br,I;$(*-!@c)]","Aromatic",0},
+  {1,"Halogen.Aliphatic","[$([F,Cl,Br,I]-!@C);!$([F,Cl,Br,I]-!@C-!@[F,Cl,Br,I])]","Aliphatic",0},
+  {1,"Halogen.NotFluorine","[$([Cl,Br,I]-!@[#6]);!$([Cl,Br,I]-!@C-!@[F,Cl,Br,I]);!$([Cl,Br,I]-[C,S](=[O,S,N]))]","Not Fluorine",0},
+  {2,"Halogen.NotFluorine.Aliphatic","[$([Cl,Br,I]-!@C);!$([Cl,Br,I]-!@C-!@[F,Cl,Br,I]);!$([Cl,Br,I]-[C,S](=[O,S,N]))]","Aliphatic Not Fluorine",0},
+  {2,"Halogen.NotFluorine.Aromatic","[$([Cl,Br,I]-!@c)]","Aromatic Not Fluorine",0},
+  {1,"Halogen.Bromine","[Br;$([Br]-!@[#6]);!$([Br]-!@C-!@[F,Cl,Br,I]);!$([Br]-[C,S](=[O,S,N]))]","Bromine",0},
+  {2,"Halogen.Bromine.Aliphatic","[Br;$(Br-!@C);!$(Br-!@C-!@[F,Cl,Br,I]);!$(Br-[C,S](=[O,S,N]))]","Aliphatic Bromine",0},
+  {2,"Halogen.Bromine.Aromatic","[Br;$(Br-!@c)]","Aromatic Bromine",0},
+  {2,"Halogen.Bromine.BromoKetone","[Br;$(Br-[CH2]-C(=O)-[#6])]","Bromoketone",0},
+
+  {0,"Azide","[N;H0;$(N-[#6]);D2]=[N;D2]=[N;D1]","Azide",0},
+  {1,"Azide.Aromatic","[N;H0;$(N-c);D2]=[N;D2]=[N;D1]","Aromatic Azide",0},
+  {1,"Azide.Aliphatic","[N;H0;$(N-C);D2]=[N;D2]=[N;D1]","Aliphatic Azide",0},
+  
+  {0,"Nitro","[N;H0;$(N-[#6]);D3](=[O;D1])~[O;D1]","Nitro",0},
+  {1,"Nitro.Aromatic","[N;H0;$(N-c);D3](=[O;D1])~[O;D1]","Aromatic Nitro",0},
+  {1,"Nitro.Aliphatic","[N;H0;$(N-C);D3](=[O;D1])~[O;D1]","Aliphatic Nitro",0},
+  
+  {0,"TerminalAlkyne","[C;$(C#[CH])]","Terminal Alkyne",0}
+};
+
+const unsigned int NUM_FUNCS =
+    static_cast<unsigned int>(sizeof(FuncDataArray) / sizeof(FuncData_t));
+
+FilterCatalog &hierarchy_get() {
+  static FilterCatalog fgroup;
+  return fgroup;
+}
+
+std::map<std::string, ROMOL_SPTR> &flatten_get() {
+  static std::map<std::string, ROMOL_SPTR> flattenedHierarchy;
+  return flattenedHierarchy;
+}
+
+void hierarchy_create() {
+  FilterCatalog & fgroupHierarchy = hierarchy_get();
+  std::map<std::string, ROMOL_SPTR> &flattenedHierarchy = flatten_get();
+  
+  std::vector<FilterHierarchyMatcher*>  toplevel;
+  std::vector<FilterHierarchyMatcher*>  stack(4); // max tree depth currently 4
+  
+  for (size_t i=0; i<NUM_FUNCS; ++i) {
+    // Make a new node
+    ROMOL_SPTR pattern(SmartsToMol(FuncDataArray[i].smarts,0, true));
+    pattern->setProp("Label", FuncDataArray[i].label);
+    if (FuncDataArray[i].removalReaction) {
+      pattern->setProp("RemovalReaction", FuncDataArray[i].removalReaction);
+    }
+    flattenedHierarchy[FuncDataArray[i].name] = pattern;
+    FilterHierarchyMatcher node(SmartsMatcher(FuncDataArray[i].name,
+                                              pattern));
+
+    if (FuncDataArray[i].level == 0) {
+      toplevel.push_back(new FilterHierarchyMatcher(node));
+      stack[0] = toplevel.back();
+      
+    } else {
+      boost::shared_ptr<FilterHierarchyMatcher> real_node = \
```

Comment:
Since the size of stack is hardcoded, here we should be testing that level<4 and generating a ValueError otherwise.